--- src/main/java/uniregistrar/openapi/model/DidUrlState.java	2024-04-09 18:30:27.916587018 +0200
+++ src/main/java/uniregistrar/openapi/model/DidUrlState.java.changed	2024-04-09 18:31:19.292781832 +0200
@@ -63,7 +63,7 @@
   @JsonSubTypes.Type(value = DidUrlStateWait.class, name = "wait"),
 })
 
-public class DidUrlState extends HashMap<String, Object> {
+public class DidUrlState {
   public static final String JSON_PROPERTY_STATE = "state";
   private String state;
 
