--- src/main/java/uniregistrar/openapi/model/RequestOptions.java	2024-04-09 18:30:37.128621920 +0200
+++ src/main/java/uniregistrar/openapi/model/RequestOptions.java.changed	2024-04-09 18:31:24.856802953 +0200
@@ -49,7 +49,7 @@
   RequestOptions.JSON_PROPERTY_REQUEST_VERIFICATION_METHOD
 })
 @jakarta.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", comments = "Generator version: 7.4.0")
-public class RequestOptions extends HashMap<String, Object> {
+public class RequestOptions {
   public static final String JSON_PROPERTY_CLIENT_SECRET_MODE = "clientSecretMode";
   private Boolean clientSecretMode;
 
