--- src/main/java/uniregistrar/openapi/model/Secret.java	2024-04-09 18:30:01.052485315 +0200
+++ src/main/java/uniregistrar/openapi/model/Secret.java.changed	2024-04-09 18:31:33.228834742 +0200
@@ -46,7 +46,7 @@
   Secret.JSON_PROPERTY_VERIFICATION_METHOD
 })
 @jakarta.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", comments = "Generator version: 7.4.0")
-public class Secret extends HashMap<String, Object> {
+public class Secret {
   public static final String JSON_PROPERTY_VERIFICATION_METHOD = "verificationMethod";
   private List<SecretVerificationMethodInner> verificationMethod;
 
