// jsonld-signatures/dist/node6/lib/suites/GraphSignature2012.js
/*
 * Copyright (c) 2017-2018 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const constants = require('../constants');
const LinkedDataSignature2015 = require('./LinkedDataSignature2015');

module.exports = class GraphSignature2012 extends LinkedDataSignature2015 {
  constructor(injector) {
    let algorithm = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'GraphSignature2012';

    super(injector, algorithm);
  }

  canonize(input, options) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const jsonld = _this.injector.use('jsonld');
      const opts = {
        algorithm: 'URGNA2012',
        format: 'application/n-quads',
        expansionMap: options.expansionMap
      };
      if (options.documentLoader) {
        opts.documentLoader = options.documentLoader;
      }
      return jsonld.canonize(input, opts);
    })();
  }

  createVerifyData(input, options) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      // TODO: frame before getting signature, not just compact? considerations:
      // should the assumption be (for this library) that the signature is on
      // the top-level object and thus framing is unnecessary?

      const jsonld = _this2.injector.use('jsonld');
      const opts = { expansionMap: options.expansionMap };
      if (options.documentLoader) {
        opts.documentLoader = options.documentLoader;
      }
      const compacted = yield jsonld.compact(input, constants.SECURITY_CONTEXT_URL, opts);

      // TODO: will need to preserve `signature` when chained signature
      // option is set in the future

      // delete the existing signature(s) prior to canonicalization
      delete compacted.signature;

      const c14n = yield _this2.canonize(compacted, options);

      let verifyData = '';
      if (options.nonce !== null && options.nonce !== undefined) {
        verifyData += options.nonce;
      }
      verifyData += options.date;
      verifyData += c14n;
      if (options.domain !== null && options.domain !== undefined) {
        verifyData += '@' + options.domain;
      }
      return {
        data: verifyData,
        encoding: 'utf8'
      };
    })();
  }

  validateKey(key, options) {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      if (typeof key.publicKeyPem !== 'string') {
        throw new TypeError('Unknown public key encoding. Public key encoding must be ' + '"publicKeyPem".');
      }
      const jsonld = _this3.injector.use('jsonld');
      if (!jsonld.hasValue(key, 'type', 'CryptographicKey')) {
        throw new TypeError('Invalid key type. Key type must be "CryptographicKey".');
      }
    })();
  }
};
